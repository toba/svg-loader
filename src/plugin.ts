import SVGO from 'svgo';
import fs from 'fs';
import path from 'path';
import { Encoding } from '@toba/tools';
import { Compiler } from 'webpack';
import {
   default as HtmlWebpackPlugin,
   HtmlTagObject
} from 'html-webpack-plugin';

const name = 'html-webpack-inline-svg';

interface SvgPluginOptions {
   /**
    * SVG file paths relative to the Webpack context.
    */
   files: string[];
   svgo?: SVGO.Options;
}

const slugify = (name: string) => {
   if (name.includes('/')) {
      name = name.substr(name.lastIndexOf('/') + 1);
   }
   return name.replace('.svg', '').replace(/\s/, '');
};

/**
 * @see https://medium.com/webpack/webpack-4-migration-guide-for-plugins-loaders-20a79b927202
 * @see https://blog.johnnyreilly.com/2018/01/finding-webpack-4-use-map.html
 * @see https://github.com/DustinJackson/html-webpack-inline-source-plugin
 */
export class HtmlSvgPlugin {
   options: SvgPluginOptions;

   constructor(options?: SvgPluginOptions) {
      this.options = options === undefined ? { files: [] } : options;
   }

   apply(compiler: Compiler): void {
      /**
       * ID attribute will be updated to match each SVG filename.
       */
      const addID = { id: '' };
      /**
       * Plugin configurations.
       * @see https://github.com/svg/svgo#what-it-can-do
       */
      const plugins: { [key: string]: string | boolean | object } = {
         addAttributesToSVGElement: { attributes: [addID] },
         removeXMLNS: true,
         sortAttrs: true,
         removeViewBox: false,
         removeDimensions: true
      };
      const svgo = new SVGO({
         plugins: Object.keys(plugins).map(
            key => ({ [key]: plugins[key] } as any)
         )
      });

      compiler.hooks.compilation.tap(name, compilation => {
         if (this.options.files.length == 0) {
            return;
         }
         const options = compilation.compiler.options;
         const context: string =
            options.context !== undefined ? options.context : __dirname;

         HtmlWebpackPlugin.getHooks(compilation).alterAssetTagGroups.tapAsync(
            name,
            async (data, cb) => {
               const optimizers: Promise<any>[] = [];
               const files = this.options.files.reduce((map, name) => {
                  map.set(
                     slugify(name),
                     fs.readFileSync(path.resolve(context, name), {
                        encoding: Encoding.UTF8
                     })
                  );
                  return map;
               }, new Map<string, string>());

               files.forEach((text, id) => {
                  addID.id = id;
                  optimizers.push(svgo.optimize(text));
               });

               const svgs = await Promise.all(optimizers);

               const tag: HtmlTagObject = {
                  tagName: 'div',
                  attributes: {},
                  innerHTML: svgs.map(svg => svg.data).join(),
                  voidTag: false
               };

               data.bodyTags.push(tag);

               cb(null, data);
            }
         );
      });
   }
}
