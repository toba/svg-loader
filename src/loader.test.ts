import '@toba/test';
import { Encoding } from '@toba/tools';
import path from 'path';
import webpack from 'webpack';
import memoryfs from 'memory-fs';
import HtmlWebpackPlugin from 'html-webpack-plugin';
import { HtmlSvgPlugin } from './plugin';

const memFS = new memoryfs();

/**
 * Test loader.
 * @see https://webpack.js.org/contribute/writing-a-loader/#testing
 */
export function compile(): Promise<webpack.Stats> {
   const compiler = webpack({
      context: __dirname,
      mode: 'development',
      entry: './__mocks__/entry.js',
      output: {
         path: path.resolve(__dirname),
         filename: 'bundle.js'
      },
      // module: {
      //    rules: [
      //       {
      //          test: /\.svg$/,
      //          use: {
      //             loader: path.resolve(__dirname, '.', 'loader.ts')
      //          }
      //       }
      //    ]
      // },
      plugins: [new HtmlWebpackPlugin(), new HtmlSvgPlugin()]
   });

   compiler.outputFileSystem = memFS;

   return new Promise((resolve, reject) => {
      compiler.run((err, stats) => {
         if (err !== null) {
            reject(err);
         } else if (stats.hasErrors()) {
            reject(stats.compilation.errors[0]);
         } else {
            resolve(stats);
         }
      });
   });
}

test('works', async done => {
   const stats = await compile();

   expect(stats).toBeDefined();

   const htmlFile = path.resolve(__dirname, 'index.html');
   memFS.readFile(htmlFile, Encoding.UTF8, (err, data) => {
      expect(err).toBeNull();
      expect(data).toMatchSnapshot();
      done();
   });
});
