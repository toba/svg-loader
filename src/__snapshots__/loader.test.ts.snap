// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`works 1`] = `
"<!DOCTYPE html>
<html>
  <head>
    <meta charset=\\"utf-8\\">
    <title>Webpack App</title>
  <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\"></head>
  <body>
  <script src=\\"bundle.js\\"></script></body>
</html>"
`;
